### PR TITLE
[FLINK-2626][CORE]add averageaccumulator

### PR DESCRIPTION
**Pull request by HuangWHWHW**
Add class averageaccumulator for FLINK-accumulator.
And test for it.

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/accumulators/AverageAccumulator.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+		this.count++;
+		this.localValue += value;
+	}
+
+	public void add(double value) {
+		this.count++;
+		this.localValue += value;
+	}
+
+	public void add(Long value) {
```
Comment (sachingoel0101) [inline flink-core/src/main/java/org/apache/flink/api/common/accumulators/AverageAccumulator.java]: Why do we need this? `add(long)` should suffice.

File: flink-core/src/main/java/org/apache/flink/api/common/accumulators/AverageAccumulator.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		this.count++;
+		this.localValue += (double)value;
+	}
+
+	public void add(long value) {
+		this.count++;
+		this.localValue += value;
+	}
+
+	public void add(Integer value) {
```
Comment (sachingoel0101) [inline flink-core/src/main/java/org/apache/flink/api/common/accumulators/AverageAccumulator.java]: Same as above.

File: flink-core/src/main/java/org/apache/flink/api/common/accumulators/AverageAccumulator.java (hunk 1 above)
Comment (HuangWHWHW) [inline flink-core/src/main/java/org/apache/flink/api/common/accumulators/AverageAccumulator.java]: I`m not quite sure.
Just the same as the class LongCounter .
You can see the code following:

public class LongCounter implements SimpleAccumulator<Long> {

	private static final long serialVersionUID = 1L;
	
	private long localValue = 0;

	/**
	 * Consider using {@link #add(long)} instead for primitive long values
	 */
	@Override
	public void add(Long value) {
		this.localValue += value;
	}

	public void add(long value){
		this.localValue += value;
	}

	@Override
	public Long getLocalValue() {
		return this.localValue;
	}
	
	@Override
	public void merge(Accumulator<Long, Long> other) {
		this.localValue += other.getLocalValue();
	}
	
	@Override
	public void resetLocal() {
		this.localValue = 0;
	}

	@Override
	public LongCounter clone() {
		LongCounter result = new LongCounter();
		result.localValue = localValue;
		return result;
	}

	@Override
	public String toString() {
		return "LongCounter " + this.localValue;
	}
}

Comment (sachingoel0101) [inline flink-core/src/main/java/org/apache/flink/api/common/accumulators/AverageAccumulator.java]: That's required by the interface, just as `add(Double)` is required here.

**Comment by HuangWHWHW** [comment]
@sachingoel0101 
Hi,
I removed these two methods as you said.

**Comment by StephanEwen** [comment]
Looks good, +1 to merge this

**Comment by mjsax** [comment]
Why is `public void add(Double value)` overloaded with `public void add(double value)`? The usage of primitive types and there class counterparts seem to be mixed... Why?

**Comment by HuangWHWHW** [comment]
@mjsax 
Do you mean that remove the `public void add(double value)`?
I`m not quite sure how this can be used in the feature.

@StephanEwen 
Thank you!
Otherwise, Can you take a look about my earlier PRs?
Since I have been fixing those for a long time and recently I took fixes and passed the CI of them.
Many thanks for you and sorry for wasting your time.

**Comment by sachingoel0101** [comment]
@mjsax , primitive add methods for accumulators were introduced in #942 to get rid of potential unnecessary boxing and unboxing.

**Comment by mjsax** [comment]
Yes, I don't see any advantage in having this overload. I would also change `add(long value)` and `add(int value)` into `add(Long value)` and `add(Integer value)` for consistency (I actually prefer primitive types, too; however, for this case class types seems to be more appropriate because the inherited methods use `Double` as parameter and as return type in `getLocalValue()`).

 But you should wait for @StephanEwen feedback before chancing anything. I am new to accumulators...

File: flink-core/src/test/java/org/apache/flink/api/common/accumulators/AverageAccumulatorTest.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+package org.apache.flink.api.common.accumulators;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/**
```
Comment (mjsax) [inline flink-core/src/test/java/org/apache/flink/api/common/accumulators/AverageAccumulatorTest.java]: This is obvious. Please remove comment.

**Comment by sachingoel0101** [comment]
@mjsax, I'm not sure what's the need for those. The `add(Double)` is only needed because the interface requires this. `add(Long)` is equivalent to `add(long)` after unboxing, which is exactly what would happen inside the implementation of `add(Long)`.

File: flink-core/src/test/java/org/apache/flink/api/common/accumulators/AverageAccumulatorTest.java (hunk 3 above)
Comment (HuangWHWHW) [inline flink-core/src/test/java/org/apache/flink/api/common/accumulators/AverageAccumulatorTest.java]: Sorry...comment?
You mean this class documentation----'/*XXX*/'?

Comment (mjsax) [inline flink-core/src/test/java/org/apache/flink/api/common/accumulators/AverageAccumulatorTest.java]: Yes.

**Comment by mjsax** [comment]
Fail enough. Ignore my comment.

File: flink-core/src/test/java/org/apache/flink/api/common/accumulators/AverageAccumulatorTest.java (hunk 3 above)
Comment (HuangWHWHW) [inline flink-core/src/test/java/org/apache/flink/api/common/accumulators/AverageAccumulatorTest.java]: Ok, I'll remove it.

**Comment by HuangWHWHW** [comment]
@sachingoel0101 
@mjsax 
Otherwise, could you tell why the CI is failed?

**Comment by mjsax** [comment]
```
Tests in error:
SlotCountExceedingParallelismTest.testNoSlotSharingAndBlockingResult » JobExecution
```

and 

```
Failed tests:
KafkaITCase>KafkaTestBase.prepare:155 Test setup failed: Unable to connect to zookeeper server within timeout: 6000
KafkaProducerITCase>KafkaTestBase.prepare:155 Test setup failed: Unable to connect to zookeeper server within timeout: 6000

Tests in error:
KafkaITCase>KafkaTestBase.shutDownServices:196 » NullPointer
KafkaProducerITCase>KafkaTestBase.shutDownServices:196 » NullPointer
```

**Comment by HuangWHWHW** [comment]
@mjsax 
Thank you for the info.
I take a fix about your comment and rerun the CI.
As for the error, I have no idea.
Is there any bugs in other module?

**Comment by HuangWHWHW** [comment]
@sachingoel0101 
Thank you for spending time in explaining.

**Comment by mjsax** [comment]
To me, the tests seems to be unstable. Maybe we need to open a JIRA for them. I will report them on the dev mailing list.

**Comment by HuangWHWHW** [comment]
@mjsax 
Otherwise, since this accumulator is prepared for the cust metrics in storm-compatibility.
Can you give me a storm example in using the cust metrics?
Or is it more suitable to send a email to you for this question on dev mailing list?

**Comment by mjsax** [comment]
We should discuss this over the dev list.

**Comment by StephanEwen** [comment]
I am pushing a hotfix to increase timeouts in the Kafka tests. That will hopefully make the Kafka tests more reliable on Travis

**Comment by HuangWHWHW** [comment]
@StephanEwen 
Just a minor thing:
Does the zk do re-connection if it failed to set up a node or cannot receive the msg from any node?
As I know the zk is unstable in some cases and I have no idea about what cases.

**Comment by StephanEwen** [comment]
I think you can cinfigure ZooKeeper to do that.

The problem in that test was that the Kafka server has an initial timeout when trying to connect to ZooKeeper. Here, ZooKeeper is an embedded mini-cluster-zookeeper for tests. Apparently either that mini cluster did not start properly, or the build was accidentally very slow (it happens in these shared VM environments) so that the timeout triggered.

**Comment by HuangWHWHW** [comment]
@StephanEwen 
Thank you.
I'll take learning in your further hotfix.

**Comment by mjsax** [comment]
+1 for merging.

**Pull request closed without merging**